lib/lseek64_stubs.c

```diff
@@ -46,12 +46,10 @@ CAMLprim value stub_lseek64_data(value fd, value ofs) {
   off_t c_ret;
 
   caml_enter_blocking_section();
-#if defined(SEEK_DATA)
   c_ret = lseek(c_fd, c_ofs, SEEK_DATA);
-#else
-  /* Set the file pointer to ofs; pretend there is data */
-  c_ret = lseek(c_fd, c_ofs, SEEK_SET);
-#endif
+  /* If there is no SEEK_DATA support, 
+     set the file pointer to ofs; pretend there is data */
+  if (c_ret == -1) c_ret = lseek(c_fd, c_ofs, SEEK_SET);
```

Comment:
In fact I think you can assume that EINVAL when doing SEEK_HOLE/SEEK_DATA does mean that the filesystem does not support sparse files.   The manual page says that errno is set to EINVAL in the following situations:

 1. whence is invalid 
 2. the resulting file offset would be negative
 3. the resulting offset would be beyond the end of a seekable device

I think when you are doing SEEK_HOLE/SEEK_DATA on a normal filesystem, only case 1 can happen.  You can't seek backwards for a hole, so situation 2 can't occur, and unless the file is on a tape drive you are not going to run off the end of the device, so situation 3 is also out.

Nevertheless, @andyhhp's comment still stands.   This patch currently re-tries the seek after any error, not just in the case where the filesystem does not support sparse files.